Instantaneous BW and IOPS reporting to Prometheus

aaronlee121 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Added instantaneous BW and IOPS for Prometheus to collect when called. Current implementation is by having a go routine that will block on a channel until Prometheus collector is called, then signal to IO threads to record IO. The metrics thread will wait 1 second then signal to stop recording and report the number of IO that happened in that 1 second to Prometheus collector.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
For management plane to see bandwidth and IOPS; we have no implementation of this currently.

## Regression
<!-- Is this PR fixing a regression? (Yes / No) -->
<!-- If Yes, optionally please include minio version or commit id or PR# that caused this regression, if you have these details. -->
No

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Ran Elbencho and s3-benchmark then signaled Prometheus to collect and compared Prometheus output to reported BW/IOPS. Mostly relied on Elbencho for this, since it has live output feature that makes it easier to compare real-time stats.

**Example Elbencho vs Promtheus output**
Elbencho:
```
Phase: READ  CPU:  52%  Active: 32  Elapsed: 10s
 Rank   %    DoneMiB      MiB/s       IOPS      Files    Files/s
Total   7      12414       1301       1301      12414       1301
```
Prometheus BW/IOPS reported at about the same time:
```
# HELP minio_metrics_bw Current BW of this Minio instance (KiB/s)
# TYPE minio_metrics_bw counter
minio_metrics_bw 1.32096e+06
# HELP minio_metrics_iops Current IOPS of this Minio instance
# TYPE minio_metrics_iops counter
minio_metrics_iops 1290
```

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

aaronlee121 committed:
Added IOPS/BW reporting to Prometheus

Integrated instantaneous IOPS/BW collection by recording 1 second of IO and outputting it to Prometheus collector.
Removed one prometheus minio collector from registering in metrics.go because it was causing Collect to run twice, didn't see any issues with doing it

aaronlee121 committed:
Merge branch 'master' of https://github.com/OpenMPDK/dss-minio into bw_iops_reporting

aaronlee121 committed:
Change BW units to bytes/s for management plane

Management plane asked for BW to reported in bytes/s, removing the conversion to KB

aaronlee121 committed:
Removed KiB/s label from Prometheus output, using bytes instead now

somnathr commented on cmd/globals.go:
```diff
@@ -190,6 +190,10 @@ var (
+		globalCurrBW uint64
+		globalMetricsChan = make(chan uint32)
+		globalCollectMetrics uint32
 
```
Fix the alignment if it is really not aligned to existing, use spaces rather than tabs if needed

aaronlee121 replied:
For some reason, my editor started using tabs instead of spaces and I couldn't tell since they look the same. I'll fix all the indentation in a separate commit so it doesn't show up with the implementation changes

somnathr commented on cmd/metrics.go:
```diff
@@ -36,9 +37,11 @@ var (
 	prometheus.MustRegister(httpRequestsDuration)
-	prometheus.MustRegister(newMinioCollector())
+	//prometheus.MustRegister(newMinioCollector())
 }
```
Remove commented

aaronlee121 replied:
Removed this with indentation changes

somnathr commented on cmd/object-handlers.go:
```diff
@@ -1419,6 +1423,11 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
+		atomic.AddUint64(&globalCurrBW, uint64(objInfo.Size))
+		atomic.AddUint64(&globalCurrIOCount, 1)
+	}
+
```
I think we need to separate IO and BW counters for PUT/GET, because GUI needs to show PUT/GET BW/IOPs separately.
Lets add counters for Del as well, no need for BW counters for DEL. Also, there are other calls in this file like Headobject etc that is also considered for Get calls..See all other related calls to categorize on Get/Put.

somnathr commented on cmd/xl-sets.go:
```diff
@@ -623,6 +642,8 @@ func newXLSets(endpoints EndpointList, format *formatXLV3, setCount int, drivesP
         }
 
+		go metricsReporting()
+
```
indentation

aaronlee121 replied:
Fixed in separate commit

somnathr commented on cmd/metrics.go:
```diff
@@ -64,6 +67,30 @@ func (c *minioCollector) Describe(ch chan<- *prometheus.Desc) {
+	// Reset counters once reported
+	atomic.StoreUint64(&globalCurrIOCount, 0)
+	atomic.StoreUint64(&globalCurrBW, 0)
+	
```
Lets discuss, the reason I have an issue with this design as metric collection is triggered by this function. Now, metrics should be independent to Prometheus..For example, if there is a request to get metrics via ustat, then you need to change the design..Rather what I suggested earlier from the  metricsReporting() thread, you reset the counter every 1 sec and keep on collecting if a global variable like metric_collection is enabled during startup...The downside of that approach is, we keep on collecting the metrics till lifetime of minio..Lets discuss more..

somnathr reviewed (commented):
Lets discuss and address my comments..

aaronlee121 committed:
Fixed indentation/comments

aaronlee121 committed:
Fixed metrics collection design to constantly run and reset from thread; added new counters

Previously the design of metrics collection made it so that metrics could only be reported through Prometheus, since the code to reset counters was inside Prometheus
Now, counters are reset inside metrics collection thread while it is constantly running.
To address the issue of metrics being collected when counters are reset, new variables will keep track of last second of metrics which Prometheus will take.
Also added new counters to separate PUT, GET, and DEL BW/IOPS and now incrementing these counters in other object API handlers that do IO
NOTE: Not supporting multipart for now

aaronlee121 committed:
Merge branch 'master' of https://github.com/OpenMPDK/dss-minio into bw_iops_reporting

somnathr reviewed (approved):
Looks good
